sdk/program/src/bignum.rs

```diff
@@ -0,0 +1,851 @@
+                    lhs_ffi_in_addr: *const u64,
+                    rhs_ffi_in_addr: *const u64,
+                    bn_syscall_result: *mut u64,
+                ) -> u64;
+            }
+            let lhs_ffi_in = {
+                FfiBigNumber {
+                    data: self.value.as_ptr() as *const _ as u64,
+                    data_len: self.value.len() as u64,
+                    is_negative: self.is_negative,
+                }
+            };
+            let rhs_ffi_in = {
+                FfiBigNumber {
+                    data: modulus.value.as_ptr() as *const _ as u64,
+                    data_len: modulus.value.len() as u64,
+                    is_negative: modulus.is_negative,
+                }
+            };
+            let value_len = modulus.value.len();
+            let mut value = Vec::<u8>::with_capacity(value_len as usize);
+            let mut bn_syscall_result = {
+                FfiBigNumber {
+                    data: value.as_mut_ptr() as *mut _ as u64,
+                    data_len: value_len as u64,
+                    is_negative: false,
+                }
+            };
+            unsafe {
+                sol_bignum_mod_inv(
+                    &lhs_ffi_in as *const _ as *const u64,
+                    &rhs_ffi_in as *const _ as *const u64,
+                    &mut bn_syscall_result as *mut _ as *mut u64,
+                );
+                value.set_len(bn_syscall_result.data_len as usize);
+                Self {
+                    is_negative: bn_syscall_result.is_negative,
+                    value,
+                }
+            }
+        }
+    }
+
+    /// Log a `BigNum` from a program to the solana logs
+    pub fn log(&self) {
+        #[cfg(not(target_arch = "bpf"))]
+        {
+            crate::program_stubs::sol_log(&self.value.to_dec_str().unwrap());
+        }
+        #[cfg(target_arch = "bpf")]
+        {
+            extern "C" {
+                fn sol_bignum_log(lhs_ffi_in_addr: *const u64) -> u64;
+            }
+            let lhs_ffi_in = {
+                FfiBigNumber {
+                    data: self.value.as_ptr() as *const _ as u64,
+                    data_len: self.value.len() as u64,
+                    is_negative: self.is_negative,
+                }
+            };
+            unsafe { sol_bignum_log(&lhs_ffi_in as *const _ as *const u64) };
+        }
+    }
+}
+
+impl PartialEq for BigNumber {
+    fn eq(&self, other: &Self) -> bool {
+        #[cfg(not(target_arch = "bpf"))]
+        {
+            self.value == other.value && self.value.is_negative() == other.value.is_negative()
+        }
+        #[cfg(target_arch = "bpf")]
+        {
+            self.value == other.value && self.is_negative == other.is_negative
+        }
+    }
+}
+impl Clone for BigNumber {
+    fn clone(&self) -> Self {
+        #[cfg(not(target_arch = "bpf"))]
+        {
+            match openssl::bn::BigNum::from_slice(&self.value.to_vec()) {
```

Comment:
`BigNum` doesn't implement `Clone`?

Reply:
@jackcmay OpenSSL does not have the trait `Clone`

Reply:
Lame